finality-aleph/src/data_io.rs

```diff
@@ -1,13 +1,190 @@
-use crate::{Error, Metrics};
+use crate::{network::AlephNetworkData, Error, Metrics};
 use aleph_bft::OrderedBatch;
 use futures::channel::{mpsc, oneshot};
-use log::debug;
+use lru::LruCache;
 use parking_lot::Mutex;
 use sp_consensus::SelectChain;
 use sp_runtime::traits::{Block, Header};
-use std::{sync::Arc, time::Duration};
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    marker::PhantomData,
+    sync::Arc,
+    time::Duration,
+};
 
 const REFRESH_INTERVAL: u64 = 100;
+use futures::channel::mpsc::{UnboundedReceiver, UnboundedSender};
+use futures_timer::Delay;
+use log::{debug, trace};
+use sc_client_api::{backend::Backend, BlockchainEvents};
+use sp_runtime::generic::BlockId;
+use tokio::stream::StreamExt;
+
+type MessageId = u64;
+const AVAILABLE_BLOCKS_CACHE_SIZE: usize = 1000;
+const PENDING_MESSAGE_LIFETIME: MessageId = 100_000;
+const TIMEOUT_DURATION: Duration = Duration::from_millis(60000);
+
+pub(crate) struct DataStore<B, C, BE>
+where
+    B: Block,
+    C: crate::ClientForAleph<B, BE> + BlockchainEvents<B> + Send + Sync + 'static,
+    BE: Backend<B> + 'static,
+{
+    next_message_id: MessageId,
+    ready_messages_tx: UnboundedSender<AlephNetworkData<B>>,
+    messages_rx: UnboundedReceiver<AlephNetworkData<B>>,
+    dependent_messages: HashMap<B::Hash, HashSet<MessageId>>,
+    available_blocks: LruCache<B::Hash, ()>,
+    message_requirements: HashMap<MessageId, usize>,
+    pending_messages: HashMap<MessageId, AlephNetworkData<B>>,
+    client: Arc<C>,
+    _phantom: PhantomData<BE>,
+}
+
+impl<B, C, BE> DataStore<B, C, BE>
+where
+    B: Block,
+    C: crate::ClientForAleph<B, BE> + Send + Sync + 'static + BlockchainEvents<B>,
+    BE: Backend<B> + 'static,
+{
+    pub(crate) fn new(
+        client: Arc<C>,
+        ready_messages_tx: UnboundedSender<AlephNetworkData<B>>,
+        messages_rx: UnboundedReceiver<AlephNetworkData<B>>,
+    ) -> Self {
+        DataStore {
+            next_message_id: 0,
+            client,
+            message_requirements: HashMap::new(),
+            dependent_messages: HashMap::new(),
+            pending_messages: HashMap::new(),
+            available_blocks: LruCache::new(AVAILABLE_BLOCKS_CACHE_SIZE),
+            ready_messages_tx,
+            messages_rx,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub(crate) async fn run(&mut self, mut exit: oneshot::Receiver<()>) {
+        let mut timeout = Delay::new(TIMEOUT_DURATION);
```

Comment:
Normally I'm quite lenient when it comes to names, but I don't like timeout -- this has a negative association that something went wrong. Here this is something closer to a "periodic maintenance"

Reply:
I changed it to `PERIODIC_MAINTENANCE_INTERVAL`